File: python/251-300/282. Expression Add Operators.py
class Solution(object):

    def addOperators(self, num, target):
        """
        :type num: str
        :type target: int
        :rtype: List[str]
        """
        res = []

        def dfs(num, path, cur, last, target):
            if not num:
                print path, cur, last
            if not num:
                if cur == target:
                    res.append(path)
                return
            for i in range(1, len(num) + 1):
                val = num[:i]
                if i == 1 or (i > 1 and num[0] != "0"):
                    dfs(num[i:], path + "+" + val,
                        cur + int(val), int(val), target)
                    dfs(num[i:], path + "-" + val, cur -
                        int(val), -int(val), target)
                    dfs(num[i:], path + "*" + val, cur - last +
                        last * int(val),  last * int(val), target)
        for i in range(1, len(num) + 1):
            if i == 1 or (i > 1 and num[0] != "0"):
                dfs(num[i:], num[:i], int(num[:i]), int(num[:i]), target)
        return res


sol = Solution()
print sol.addOperators("123", 6)

File: cpp/251-300/282. Expression Add Operators.cc
class Solution {
private:
    string op = "+-*";
public:
    vector<string> addOperators(string num, int target) {
        vector<string> res;
        helper(num, 0, target, "", res);
        return res;
    }
    bool cal(string s, int target) {
        long res = 0;
        long cur_res = 0;
        char op = '+';
        for (int i = 0; i < s.size();) {
            if (isdigit(s[i])) {
                long num = 0;
                while (i < s.size() && isdigit(s[i])) {
                    num = num * 10 + s[i] - '0';
                    i++;
                }
                switch (op) {
                case '+':
                    cur_res += num; break;
                case '-':
                    cur_res -= num; break;
                case '*':
                    cur_res *= num; break;
                }
            }
            else {
                if (s[i] == '-' || s[i] == '+') {
                    res += cur_res;
                    cur_res = 0;
                }
                op = s[i++];
            }
        }
        return (res + cur_res) == target;
    }
    void helper(string num, int index, int target, string path, vector<string> &res) {
        if (index == num.size()) {
            if (cal(path, target))
                res.push_back(path);
        }
        for (int i = 1; i + index <= num.size(); i++) {
            if (i >= 2 && num[index] == '0')
                continue;
            if (path.empty())
            {
                helper(num, index + i, target, num.substr(index, i), res);
            }
            else {
                for (char c : op) {
                    helper(num, index + i, target, path + string(1, c) + num.substr(index, i), res);
                }
            }
        }
    }
};